crates/dbsp/src/trace/test_batch.rs

```diff
@@ -438,9 +609,71 @@ where
         self.index = 0;
     }
 
+    fn fast_forward_keys(&mut self) {
+        todo!()
+    }
+
     fn rewind_vals(&mut self) {
         todo!()
     }
+
+    fn step_val_reverse(&mut self) {
+        let current_key = self.data[self.index].0 .0.clone();
+        let current_val = self.data[self.index].0 .1.clone();
+
+        while self.data[self.index].0 .1 == current_val {
+            if self.index > 0 && self.data[self.index - 1].0 .0 == current_key {
+                self.index -= 1;
+            } else {
+                self.val_valid = false;
+                return;
+            }
+        }
+
+        let current_val = self.data[self.index].0 .1.clone();
+
+        while self.index > 0
+            && self.data[self.index - 1].0 .0 == current_key
+            && self.data[self.index - 1].0 .1 == current_val
+        {
+            self.index -= 1;
+        }
+    }
+
+    fn seek_val_reverse(&mut self, val: &V) {
+        loop {
+            if !self.val_valid() || self.val() >= val {
+                break;
+            }
+            self.step_val_reverse();
+        }
+    }
+
+    fn seek_val_with_reverse<P>(&mut self, _predicate: P)
+    where
+        P: Fn(&V) -> bool + Clone,
+    {
+        todo!()
+    }
+
+    fn fast_forward_vals(&mut self) {
+        let current_key = self.data[self.index].0 .0.clone();
+
+        while self.index + 1 < self.data.len() && self.data[self.index + 1].0 .0 == current_key {
+            self.index += 1;
+        }
+
+        let current_val = self.data[self.index].0 .1.clone();
+
+        while self.index > 0
+            && self.data[self.index - 1].0 .0 == current_key
+            && self.data[self.index - 1].0 .1 == current_val
```

Comment:
```suggestion
            && self.data[self.index - 1].0 == (current_key, current_val)
```

Reply:
This won't do, there's also the time field in that tuple.